edgecast/client/ec_client_test.go

```diff
@@ -0,0 +1,705 @@
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/EdgeCast/ec-sdk-go/edgecast/auth"
+	"github.com/EdgeCast/ec-sdk-go/edgecast/internal/testhelper"
+)
+
+var (
+	testLog = testhelper.GetTestLogger("ec_client_test")
+)
+
+func TestSetBody(t *testing.T) {
+	cases := []struct {
+		name     string
+		body     interface{}
+		expected string
+	}{
+		{
+			name:     "Happy Path - string body",
+			body:     "some string payload",
+			expected: "some string payload",
+		},
+		{
+			name: "Happy Path - struct body",
+			body: sampleData{
+				StringData: "some string",
+				IntData:    100,
+				Nested: nestedData{
+					ID: "abcd",
+				},
+			},
+			expected: `{"StringData":"some string","IntData":100,"Nested":{"ID":"abcd"}}`,
+		},
+	}
+	for _, c := range cases {
+		req := request{}
+		req.setBody(c.body)
+
+		var actual string
+		switch c.body.(type) {
+		case string:
+			actual = string(req.rawBody.([]byte))
+		default:
+			buf := req.rawBody.(*bytes.Buffer)
+			actual = buf.String()
+			// JSON encoding adds a newline character to the end, so trim it
+			actual = strings.TrimSuffix(actual, "\n")
+		}
+
+		if strings.Compare(c.expected, actual) != 0 {
+			t.Fatalf("%s: Expected %+v but got %+v", c.name, c.expected, actual)
+		}
+	}
+}
+
+func TestSetAuthorization(t *testing.T) {
+	cases := []struct {
+		name       string
+		token      string
+		throwError bool
+		expected   string
+	}{
+		{
+			name:       "Happy Path",
+			token:      "AUTH_TOKEN",
+			throwError: false,
+			expected:   "AUTH_TOKEN",
+		},
+		{
+			name:       "Error Path",
+			throwError: true,
+		},
+	}
+	for _, c := range cases {
+		req := request{}
+		authProvider := testAuthProvider{
+			Auth:  c.token,
+			Error: c.throwError,
+		}
+		err := req.setAuthorization(authProvider)
+		if c.throwError {
+			if err == nil {
+				t.Fatalf("Case '%s': expected an error, but got none", c.name)
+			}
+		} else {
```

Comment:
Oh, well testAuthProvider can be set up to intentionally throw an error - what we are essentially doing here is checking whether the method setAuthorization is bubbling up the error appropriately and not swallowing it. Notice the "throwError" property

Reply:
I think the case I am thinking about is if throwError is false but an error is generated in setAuthorization. The comparison would fail since there was an error but I believe in this flow the error wouldn't bubble up into the unit test failure message.

Reply:
@wdawg33 oh gotcha, yes absolutely we should be checking that. You're exactly right about that case not being possible here, which is why I omitted the check. For example, if you check `TestSendRequest` I included a check there.
I'll add a comment just in case it's unclear to anybody reading the code as well.

Reply:
Actually, I will add the check anyway in case the code ever gets modified unexpectedly... and add a comment saying it should never happen.